mlflow/store/rest_store.py

```diff
@@ -38,27 +39,37 @@ def _api_method_to_info():
 _METHOD_TO_INFO = _api_method_to_info()
 
 
+class RestException(Exception):
+    """Exception thrown on 400-level errors from the REST API"""
+    def __init__(self, json):
+        message = json['error_code']
+        if 'message' in json:
+            message = "%s: %s" % (message, json['message'])
+        super(RestException, self).__init__(message)
+        self.json = json
+
+
 class RestStore(AbstractStore):
     """ Client for a remote tracking server accessed via REST API calls """
 
-    def __init__(self, hostname):
+    def __init__(self, http_request_params):
         super(RestStore, self).__init__()
-        self.hostname = hostname
-
-    def _get_headers(self):  # noqa
-        """ Returns header for REST API requests. Can be overridden in subclasses """
-        return None
-
-    def _get_auth(self):  # noqa
-        """ Returns auth for REST API requests. Can be overridden in subclasses """
-        return None
+        self.http_request_params = http_request_params
+        if not http_request_params['hostname']:
+            raise Exception('hostname must be provided to RestStore')
 
     def _call_endpoint(self, api, json_body):
         endpoint, method = _METHOD_TO_INFO[api]
         response_proto = api.Response()
-        js_dict = http_request(hostname=self.hostname, endpoint=endpoint, method=method,
-                               auth=self._get_auth(), headers=self._get_headers(),
-                               req_body_json=json_body, params=None)
+        # Convert json string to json dictionary, to pass to requests
+        if json_body:
+            json_body = json.loads(json_body)
+        js_dict = http_request(endpoint=endpoint, method=method,
+                               req_body_json=json_body, params=None, **self.http_request_params)
+
+        if 'error_code' in js_dict:
```

Comment:
Added rudimentary error_code handling here. We should rewrite this section of the code some time; this was just to avoid it from being deserialized as the expected Response object.

Reply:
Makes sense. Orthogonally: realized we need to fix error-handling in the open-source server (so that it also returns a well-formed JSON error response instead of a 500 internal error for invalid requests). For example try running `mlflow ui` and hitting `http://localhost:5000/api/2.0/preview/mlflow/experiments/get?experiment_id=some_invalid_id`. Will make a GitHub issue to track this

Reply:
Oh, got it, yeah. They will definitely need to match error behaviors of the UI. We should make sure to do this in the unit/integration test @tomasatdatabricks